Comprehensive theming

singingwolfboy opened the pull request:
@lamagnifica Here's my first stab at writing some documentation for the comprehensive theming feature. *Before* you start correcting grammar and wording, can you go over this at a high level? We should figure out:

* Is this the right place in the documentation for this information? Should it be broken out into multiple places?
* Who will the audience be? Does the tone and level of detail fit the audience? (There will probably be multiple different audiences interested in this.)
* From the point of view of someone who does not already know this information, what information is missing?

lamagnifica commented:
Thanks, @singingwolfboy . I'm new to the project yet, still learning about what we have and what's planned, but here's a very early developmental edit. 

I propose an entirely new section of the Installing, Configuring, Running guide, to follow the current "Configuring" section (which is getting overly bloated).

In a new directory under install_operations/source, set up an index.rst file with overview info, and then reference a set of other files, each of which will contain either:
- one concept: a thing that a specific audience needs to know about this feature (an example is the info in the  "Definitions" section, although currently is is presented as reference)
- one procedure: the steps that a specific audience will need to complete to achieve a goal

So, from what I understand so far, the audience(s) are most likely to be sys admins, who need to: 
—> receive files (templates, images, whatever) from designers *
—> make sure those assets are in all ways correct (the right name, no size problems, in the right language, whatevs)
—> define a place to put the files and then put the files in the right place (your "Creating a Theme")
—> tell the system where that right place is (your "Activating a Theme")
—> test it (would they use devstack?)
—> put it in production
—> cope with platform changes and upgrades in the future 
… and/or for some subset of this audience, 
—> adjust their current Stanford or microsite theme in some way to deal with this new structure

* there is also maybe an audience of designers. For example, if there is are requirements that different types of assets will need to meet (like, the logo should be bigger than this but smaller than that, and in png or jpg format only, but a default profile image should have entirely different dimensions and has to be in gif format only, ad infinitum), then putting all of that in a reference, that the sys admins can just point to, would probably make sense, since it's the designers who can do something about it (though the sys admins also need access to it if it helps them troubleshoot theme changes that don't work)

One observation: the "Background" section that you have now would be very well suited for a blog post or a release notes-type announcement. Typically, though, we don't spend a lot of real estate in the guides describing the past or the future, but instead keep a tight focus on what the capabilities are in the now. The docs help people solve specific problems that they have now.

I also have some other thought, suggestions, and questions that we can discuss when we meet, like, 
- a theme isn't really just a directory structure, it's a collection of assets that are placed into a directory structure, so maybe some word choices need to get tweaked
- it isn't clear to me but I have a feeling that any given Open edX site can only have one theme, so maybe instead of Creating "a" or Activating "a" we'd steer toward Creating/Activating "your"…

singingwolfboy commented:
@lamagnifica: I've reworked this documentation, and I think it's ready for another review. Can you take a look?

@talbs: I added a section on Sass theming, and since I wasn't sure which Sass variables the UX team is prepared to support (if any), I wrote the documentation to say that none of them were. However, I included information on where enterprising theming-turbos could find these unsupported Sass variables, should they choose to use them anyway. Could you review this section, and let me know if what I've written is reasonable? (I figure that the UX team can work on picking a few Sass variables to support for theming purposes, but that's not necessary for the first iteration of this feature.)

lamagnifica commented on en_us/install_operations/source/theming/index.rst:
```diff
@@ -0,0 +1,20 @@
+######################################################
+
+Almost everyone who runs Open edX wants to change the way it looks and feels,
+to suit their unique website and courses. This section describes how to create
```
suit --> brand (?) uniquely identify (?)

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,68 @@
+
+Any application, including Open edX, can be loosely divided into two parts:
+the user interface ("how it looks"), and the application logic ("how it works").
+For example, when considering a multiple-choice problem in Open edX, the
```
no hyphen in multiple choice

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,68 @@
+Theming consists of changing the user interface without changing the application
+logic. When setting up an Open edX website, the website operator will often
+want to set up their own logo, modify the color scheme, change links in the
+header and footer for SEO purposes, and so on. However, although the user
```
spell out what SEO stands for

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,68 @@
+header and footer for SEO purposes, and so on. However, although the user
+interface may look different, the application logic should remain the same,
+so that Open edX continues to work properly, teaching amazing courses to students
+everywhere. A well-designed theme will preserve the general layout and structure
```
strike ", teaching amazing courses to students everywhere" (true, but editorializing -- save for the blog post!)

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,68 @@
+***********
+Definitions
+***********
+
```
maybe pull this out into a separate rst "Theming Glossary" file

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+****************
+
+To make a theme, start by creating a directory with a reasonable name for
+your theme. It's a good idea to put a README file into this directory with
```
avoid contractions

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+Creating a Theme
+****************
+
+To make a theme, start by creating a directory with a reasonable name for
```
a reasonable --> an identifying

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+To make a theme, start by creating a directory with a reasonable name for
+your theme. It's a good idea to put a README file into this directory with
+a description of what your theme is for, and to put the entire directory under
+version control. In this example, we'll call the theme ``my-theme``:
```
Use complete sentences to introduce examples, rather than ending with a colon.
In this example, the theme and its directory are named ``my-theme``.

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+    my-theme
+    └── README.rst
+
+Next, create a subdirectory for the Open edX subsystem that you want your theme
```
We don't use the word "subsystem" anywhere in the documentation now. We seem to use components for what you mean. http://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/named-release-cypress/devstack/install_devstack.html#components

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+Next, create a subdirectory for the Open edX subsystem that you want your theme
+to apply to. One theme can apply to multiple subsystems: LMS, Studio,
+certificates, emails, etc. For the moment, only the LMS subsystem is
+supported, so we'll make a ``lms`` directory.
```
what does "For the moment, only the LMS subsystem is supported" mean? If it means that you can only theme the LMS in the MVP version of theming, we need to get in front of that much earlier, rather than giving the impression that you can theme the entire system. We should document only what is available at release, and not possible future extensions.

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,55 @@
+        |   └── images
+        |       └── logo.png
+        └── templates
+            └── header.html
```
Xref to the complete list of the assets and templates that can be themed when available.

lamagnifica replied:
Adding a sentence or two with a "what's next" statement might be helpful here. Maybe something like. "You can continue to add templates and assets to the ``my-theme`` directory. You can also theme CSS files using Sass. Then, you activate your theme."

singingwolfboy replied:
Addressed.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+framework for writing and organizing CSS files, and this gives theme authors
+more flexibility with how they modify styling on Open edX. Instead of replacing
+the CSS files outright, theme authors can override specific `Sass variables`_
+to modify parts of the CSS files without replacing *all* of them.
```
no italics

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+in the LMS comes from the ``lms/static/sass/lms-main.scss`` Sass file, so to
+override styling in that file, you must create a file named ``lms-main.scss``
+in your theme's ``sass`` directory.
+
```
wouldn't it be faster to duplicate the ``lms/static/sass/lms-main.scss`` file, move it here, and make modifications?

lamagnifica replied:
maybe not, looking at lines 52-53...

talbs replied:
Yeah, duplicating isn't exactly the workflow we want folks to go through.

singingwolfboy replied:
I don't think there's any change that needs to be made here. Addressed.

lamagnifica replied:
unless you @singingwolfboy and @talbs think that it is worth explicitly telling people not to duplicate.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+    you import the core template, otherwise they will not work properly.
+
+Organizing Your Sass with Partials
+----------------------------------
```
strike "Your" ?

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+process, Open edX produces multiple different CSS files, including special
+versions of these files for languages with right-to-left alphabets. In order
+to override Sass variables for all of these CSS files without needing to
+repeat your overrides multiple times, we recommend that theme authors use
```
I'd stick with second person to refer to your readers, so theme authors --> you

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+to override Sass variables for all of these CSS files without needing to
+repeat your overrides multiple times, we recommend that theme authors use
+Sass partials to define their overrides, and import those overrides wherever
+they need.
```
they need --> they are needed? I'm not sure I'm reading this correctly -- what does "they" refer to?

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+they need.
+
+In the previous example, we overrode the ``$header-bg`` variable in the
+``lms-main.scss`` file only. We should do the same thing for the
```
why should we do it twice in these two files? If this is something they'll need to do, be explicit about what is being accomplished and why.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+    @import 'lms/static/sass/lms-main-rtl';
+
+But that means we're defining the ``$header-bg`` variable in two different places.
+If we ever want to change the background color *again* in our theme, we'll need
```
no italics

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+    @import 'lms/static/sass/lms-main-rtl';
+
+.. note::
+   The Sass partial should be in a name that starts with an underscore:
```
should be --> must be (based on language of line 101)

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+   The Sass partial should be in a name that starts with an underscore:
+   ``_overrides.scss``. However, when using ``@import`` in the other Sass
+   files in your theme, you should **not** start with an underscore.
+   Consult the documentation on `Sass partials`_ for more information.
```
standard format for xrefs is 
For more information, see {exact title of whatever you link them to}.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+
+.. _Sass: http://sass-lang.com/
+.. _Sass variables: http://sass-lang.com/documentation/file.SASS_REFERENCE.html#variables_
+.. _Sass partials: http://sass-lang.com/documentation/file.SASS_REFERENCE.html#partials
```
please add the links to the edx-documentation/en_us/links/links.rst file, and then add an include statement for that file instead, like 
.. include:: ../../../../links/links.rst

singingwolfboy replied:
Why? I don't see most of these links being shared with other pages of the documentation, so moving it to `links.rst` just makes them harder to find. If you say "just do it anyway", I'll do it anyway (since this is your project and not mine), but this doesn't make sense to me.

lamagnifica replied:
We just went (mostly) through a time consuming and somewhat painful process of getting all external links across all files in the edx-documentation repo into the single links file. Having links in the rst files throughout the repo, added by multiple writers over months, ended up causing all kinds of uninteresting problems and hard-to-find discrepancies, since we assumed all too often that a link was only used in a single file, but then it wasn't, and then it was updated in one place and not another, etc. etc. 
If that isn't at all convincing, then just do it anyway.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,177 @@
+Note that you still need to import the core Sass file in your themed
+``lms-main.scss`` file; otherwise, your Sass rules will replace all of the
+core Sass rules instead of adding to them.
+
```
Again, it would be nice to provide some possible next steps here, like an xref to activating your theme.

singingwolfboy replied:
Addressed.

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,27 @@
+a good default location is ``/edx/app/themes/my-theme``.
+
+Once the theme is installed on the server, you can activate it by setting
+the ``COMP_THEME_DIR`` Ansible variable to the full path to your theme
```
to your theme --> of your theme's (?)

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,27 @@
+
+    COMP_THEME_DIR: "/edx/app/themes/my-theme"
+
+Then, you'll need to run Ansible, so that this variable is propogated to the
```
propogated --> propagated

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,27 @@
+.. code-block:: yaml
+
+    COMP_THEME_DIR: "/edx/app/themes/my-theme"
+
```
where ``my-theme`` is the name of the directory...

singingwolfboy replied:
Addressed.

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,27 @@
+you'll need to alter those scripts to install your theme onto the server,
+as well. It's up to you where to install the theme on the server, but
+a good default location is ``/edx/app/themes/my-theme``.
+
```
I'd suggest how/where to install the theme in order to test it out first, followed by how where to install it for production.

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,27 @@
+
+.. code-block:: bash
+
+    sudo /edx/bin/update edx-platform master
```
what might I expect to happen after I do this? will I be able to see my theme immediately, or do I have to clear the browser cache or log out and back in, or...? should I click around to test things I know have been themed? etc.

singingwolfboy replied:
Addressed.

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,33 @@
+*************************
+Supported Themable Assets
+*************************
```
we should xref to this file and the other supported-{type} files from the what _is and create_ files.

singingwolfboy replied:
Addressed.

lamagnifica commented on en_us/install_operations/source/theming/supported_templates.rst:
```diff
@@ -0,0 +1,29 @@
+
+In theory, all core templates in the LMS and Studio can be overriden by themed
+templates. In practice, most of the system assumes that core templates
+will *not* be overridden by themed templates, and you may see unexpected
```
may --> might

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,33 @@
+
+In theory, all core assets in the LMS and Studio can be overriden by themed assets.
+In practice, most of the system assumes that core assets will *not* be overridden
+by themed assets, and you may see unexpected behavior if you override certain
```
may --> might

lamagnifica commented:
@singingwolfboy this is great! thanks so much for taking on this task, you have done a really fine job. 
The one area that still needs work is that, regardless of what we plan for theming, we have to be careful to document only what is possible at release. The earlier overview sections have to set expectations accurately. Otherwise, I've tried to point out some of the standards that we're working to follow in an effort to make translation less costly, and make the doc easier to understand by people whose first language isn't English.

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,68 @@
+in the future.
+
+Theming consists of changing the user interface without changing the application
+logic. When setting up an Open edX website, the website operator will often
```
Be careful of using the future tense (throughout) and try to use present tense, active voice, as much as possible.

singingwolfboy commented:
@lamagnifica Thanks for the review! I'm addressed many of your comments, and I'm in the process of addressing the rest. When this pull request is merged, it will be possible to theme templates and assets in both the LMS and Studio, but no other parts of the system (certificates, notifier emails, etc). How and where do you recommend that I document that?

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,60 @@
+to apply to. One theme can apply to multiple components: LMS, Studio,
+certificates, emails, etc. For the moment, only the LMS component is
+supported, so we will make a ``lms`` directory.
+
```
@singingwolfboy I think that this paragraph ^^ is where we give an impression of the feature's scope that is misleading or unclear. Many people will read it (skim it) and get the impression that you can theme the LMS, Studio, certificates, emails, and other unnamed parts of the platform, and then might--or might not--see the contradictory statement that theming is only supported for the LMS (right now, the word "theming" is missing). BUT, since you can theme the logo.png in Studio, neither of these statements accurately presents information about what you can do with theming right now. 
Does that make sense or am I misunderstanding the MVP?

talbs commented on en_us/install_operations/source/theming/supported_sass.rst:
```diff
@@ -0,0 +1,12 @@
+
+There are currently no supported themable Sass variables. If you want to see
+what Sass variables *can* be overridden, look in the
+``lms/static/sass/base/_variables.scss`` file in the edx-platform repository.
```
This make a lot of sense and is accurate for our support levels now, thanks! One quick note - this doesn't mention Studio nor its Sass variables, which are different than the LMS variables. Worth mentioning and pointing to the variables partial (https://github.com/edx/edx-platform/blob/master/cms%2Fstatic%2Fsass%2F_variables.scss) for that app?

talbs commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,183 @@
+
+.. note::
+
+    When overriding variables in Sass, your variables must be defined **before**
```
We might want to make a note that while most LMS (and eventually CMS) Sass variables now have the ``!default``, this technique of overriding variables won't work on any variables that do not have that flag in our app's default Sass.

singingwolfboy replied:
I feel like the section on supported Sass variables is a better place to put that information. Any supported variable will have that flag, so they don't need to worry about it here -- only when they're diving into unsupported Sass variables.

talbs commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,183 @@
+to remember to do it twice. This becomes much more difficult to keep track of
+as the number of Sass overrides in your theme grows.
+
+Instead, we will move this definition out into a partial. Add another Sass file
```
So. Much. Yes. Thanks for giving folks this direction.

talbs commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,183 @@
+
+    // Define a new Sass rule
+    body ul li {
+        font-size: 1000px;
```
This is a huge opportunity for a fun inline Sass comment that's on brand like ``// embiggen the learning!`` :P

singingwolfboy replied:
I love the idea, but I feel like it would be confusing for someone who isn't familiar with Sass or CSS. @lamagnifica, any thoughts about this?

lamagnifica replied:
not familiar with Sass or CSS enough to weigh in, but having a comment where there's likely to be a comment might be helpful

lamagnifica replied:
'cause, 1000px needs some explanation!

talbs commented on en_us/install_operations/source/theming/supported_sass.rst:
```diff
@@ -0,0 +1,25 @@
+
+There are currently no supported themable Sass variables.
+
+Note that Sass variables in the LMS and Studio *can* be overriden, but the
```
Might be worth referencing the Sass/CSS Styleguide (https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS) somewhere in this doc or in the general using Sass area. We have some details around syntax preferences as well as guidance on documenting variables (https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#colors-and-variables).

singingwolfboy replied:
I added this information to the `sass.rst` file, since it's more about *how* to use Sass overrides than about *what* can be overridden. Addressed.

talbs commented:
@singingwolfboy, thanks for this solid start to documenting theming. I left a few inline comments about additions/clarifications, but in general this looks great and is well written.

:+1:

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,59 @@
+header and footer for SEO (search engine optimization) purposes, and so on.
+However, although the user interface may look different,
+the application logic should remain the same, so that Open edX continues
+to work properly. A well-designed theme will preserve the general layout
```
I would strengthen that sentence to say that the application logic MUST remain the same (right?) along these lines:
When you make changes to the user interface, the application logic must remain the same, so that Open edX continues to work properly.

lamagnifica replied:
will preserve --> preserves

"still find it familiar" you mean, because they've already taken edX courses? or because they're taking a course that is currently running and theming changes can be disruptive? Maybe it's worth calling out some best practices for undertaking theming changes (perhaps in a follow on PR) including, do this when a minimal number of courses are running, make iterative changes to make sure learner experience isn't disrupted, ask course teams to let learners know in advance that changes to the look are coming and direct them how to give feedback (if desired)...

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,59 @@
+core templates with your own themed templates, as well as replacing some or
+all of the core assets with your own themed assets.
+
+Note that there is no limit to the number of themes you can have installed on
```
FYI a hard-to-follow rule is to only use "Note" when you are actually using the preformatted note style to draw attention to information. If it isn't worth styling as a .. note:: then rewrite. This reads more definitively if you strike "Note that " and start with "There is no limit..."

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,59 @@
+supported for theming. All templates and assets *can* be themed, but unsupported
+templates and assets may be renamed or used differently in the future, and
+these changes may occur without notice. The following pages describe the
+current supported templates and assets.
```
currently (? I think)

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,67 @@
+To make a theme, start by creating a directory with an identifying name for
+your theme. It is a good idea to put a README file into this directory with
+a description of what your theme is for, and to put the entire directory under
+version control. In this example, we will call the theme ``my-theme``.
```
avoid third person: In this example, the theme is named ``my-theme``.

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,67 @@
+
+Next, create a subdirectory for the Open edX component that you want your theme
+to apply to. One theme can apply to multiple components: LMS, Studio,
+certificates, emails, etc. For the moment, only the LMS component and the Studio
```
etc. --> and so on (avoid Latin abbreviations)

lamagnifica replied:
I think there's an assumption here that needs to be stated more explicitly:
...certificates, emails, and so on. ^You add a subdirectory for each component affected by your theme.^ For the moment, ...

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,67 @@
+For more information, see :ref:`Sass Theming` and
+:ref:`Supported Themable Sass Variables`.
+When you have finished creating your theme,
+you should :ref:`activate your theme <Activating a Theme>`.
```
should or must? I'd actually leave both of them out of this sentence.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+    overrides conform to this styleguide, but your theme may be easier to
+    maintain if it does.
+
+To override Sass variables, you should create a ``sass`` directory inside of
```
should or must? again, just leaving should out will be clear (unless they are free to name the subdir whatever they want

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+
+To override Sass variables, you should create a ``sass`` directory inside of
+the ``static`` directory in your theme. For example, if you are overriding Sass
+for the LMS, in a theme named ``my-theme``, the directory structure should look
```
should look --> looks

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+they are needed.
+
+In the previous example, we overrode the ``$header-bg`` variable in the
+``lms-main.scss`` file only. In order to make these overrides also apply when
```
strike "In order "

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+
+In the previous example, we overrode the ``$header-bg`` variable in the
+``lms-main.scss`` file only. In order to make these overrides also apply when
+Open edX users are using a right-to-left language, we should also make
```
we should also make --> you make

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+----------
+When you have finished creating your theme,
+you should :ref:`activate your theme <Activating a Theme>`.
+
```
strike should

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+Organizing Sass with Partials
+-----------------------------
+
+`Sass partials`_ are a great way of organizing Sass and reducing repetition in
```
"great" is editorializing (do not do!) however, a factual observation like, Sass partials are a widely-used way... or Many designers use Sass partials to organize... is absolutely fair game

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+
+`Sass partials`_ are a great way of organizing Sass and reducing repetition in
+your styling. The Open edX core Sass files use partials extensively, and we
+recommend that you do so, as well. During the Sass compilation
```
we recommend --> edX recommends 
(avoid 3rd person throughout)

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+the same overrides in the ``lms-main-rtl.scss`` file, as well.
+We will need another Sass file in our theme
+that matches the name of this second file that we want to override, so our
+theme directory will look like this:
```
Avoid ending sentences with colons. (they're often sentence fragments and missing an object. So you could just add "example" and a period. Here's another take, using 2nd person, using present tense, avoiding should, yadda yadda)
To override the ``lms-main-rtl.scss`` file, you add a Sass file with that name to your my-theme directory. As a result, your my-theme directory now includes the following subdirectories and files.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+                └── lms-main-rtl.scss
+
+Without using partials, the contents of the ``lms-main-rtl.scss`` file might
+look like this:
```
"...this example." or "...the following example." (throughout)

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+    // Import the core Sass file
+    @import 'lms/static/sass/lms-main-rtl';
+
+But that means we're defining the ``$header-bg`` variable in two different places.
```
avoid contractions (throughout)

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+Adding New Sass Rules
+---------------------
+In addition to overriding Sass variables in your theme, you can also define new
+Sass rules. These rules will be added to any other rules defined in the core
```
will be --> are

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,191 @@
+In addition to overriding Sass variables in your theme, you can also define new
+Sass rules. These rules will be added to any other rules defined in the core
+Sass files, rather than replacing the existing Sass rules. To add a new Sass
+rule, simply define it in the themed Sass file, or one of the Sass partials
```
simply --> you (editorializing)

singingwolfboy commented:
@lamagnifica I've addressed your feedback as much as possible. However, I have a question. You suggest avoiding the third person "we" throughout, but I find that it's very difficult to do that in the `create_theme.rst` file. That file is all about taking the reader on a step-by-step approach to creating a theme, and it's convenient to use phrasing like "Next, we'll do this." Can you suggest an alternate way of approaching this?

Also, if you have further review comments, feel free to make another review pass.

lamagnifica commented:
Yes, it takes some practice and rethinking. If you imagine yourself talking someone through a task, where they are the one sitting at the keyboard and you are looking over their shoulder, it might be easier to reorganize your approach. We aren't going to do anything, instead, the person at the keyboard is actually engaged in doing it! (Focus on the empowerment and immediacy that 2nd person and present tense give to your reader!) That way you can rephrase "Next, we'll do this" into,  To do _this thing you really want to do_, you... or Next, you complete _this step toward your goal_ .

lamagnifica commented on en_us/install_operations/source/theming/index.rst:
```diff
@@ -0,0 +1,22 @@
+
+######################################################
+Theming the Open edX Platform
+######################################################
```
per our convo:
Changing the Way Open edX Looks

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,60 @@
+the user interface ("how it looks"), and the application logic ("how it works").
+For example, when considering a multiple choice problem in Open edX, the
+user interface consists of how the options are laid out on the page, how the
+option selectors look, how the "Check Response" button looks, what sort of
```
the label is "Check" right now but there are rumors that it will be changing. What do you think about making this more generic, unlong the lines of, 
...how the option selectors look, what label appears on the button to check responses, what sort of...

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,60 @@
+to use their own logo, modify the color scheme, change links in the
+header and footer for SEO (search engine optimization) purposes, and so on.
+However, although the user interface may look different,
+the application logic must remain the same, so that Open edX continues
```
strike comma after same

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,60 @@
+to work properly. A well-designed theme preserves the general layout
+and structure of the user interface, so that users of the website still
+find it familiar and easy to use. Be careful about making sweeping changes
+to the user interface without warning: your users will be very confused!
```
+1

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,60 @@
+by the HTML files created by the templates. For example, a logo image is an
+asset. The Open edX platform already comes with a set of templates and assets
+that are used in the default user interface; these are known as "core" templates
+and "core" assets. To create a theme, you simply replace some or all of these
```
no need for the quotes

lamagnifica commented on en_us/install_operations/source/theming/what_is_theming.rst:
```diff
@@ -0,0 +1,60 @@
+to the user interface without warning: your users will be very confused!
+
+From a technical perspective, theming consists of overriding core templates
+and core assets with themed templates and themed assets. Templates are text
```
suggestion: bullet out the Templates are... and Assets are.. sentences to break up this paragraph and draw attention to these key concpets

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,71 @@
+****************
+Creating a Theme
+****************
+
```
I think I might have steered you to taking out all of the pronouns by mistake. You could sprinkle a few back in to make it clear who is performing these actions.
To make a theme, you start... (line 7)
Next, you create a... (line 17)
You add a subdirectory... (line 19)

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,71 @@
+To make a theme, start by creating a directory with an identifying name for
+your theme. It is a good idea to put a README file into this directory with
+a description of what your theme is for, and to put the entire directory under
+version control. In this example, the theme is named ``my-theme``.
```
throughout this section, I suggest putting a line feed between the sentence that introduces the example and the paragraph before.

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,71 @@
+    ├── README.rst
+    └── lms
+
+Most themes will override some assets, such as overriding the logo in the LMS.
```
strike "will", maybe replace "some" with "one or more"

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,71 @@
+        └── templates
+            └── header.html
+
+You can continue to add templates and assets to the my-theme directory.
```
``my-theme`` mono

lamagnifica commented on en_us/install_operations/source/theming/create_theme.rst:
```diff
@@ -0,0 +1,71 @@
+You can continue to add templates and assets to the my-theme directory.
+For more information, see :ref:`Supported Themable Templates`
+and :ref:`Supported Themable Assets`.
+You can also theme CSS files using Sass variables.
```
line feed before line 67

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,194 @@
+                └── lms-main-rtl.scss
+
+In this ``_overrides.scss`` file, you can define only the overrides that should
+apply to both other files, like this example.
```
both ^of the^ other

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,194 @@
+    // Make the background red
+    $header-bg: rgb(250,0,0);
+
+Now modify both other Sass files to import this partial. The contents
```
both ^of the^ other

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,194 @@
+.. note::
+   The Sass partial must be in a name that starts with an underscore:
+   ``_overrides.scss``. However, when using ``@import`` in the other Sass
+   files in your theme, you should **not** start with an underscore.
```
"you should **not** start with an underscore." doesn't make the object clear. Is it a file name that should not have the underscore?

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,194 @@
+``lms-main.scss`` file; otherwise, your Sass rules will replace all of the
+core Sass rules instead of adding to them.
+
+Next Steps
```
I don't think you need to literally include this header, or, if you don, nice to have in every file.

singingwolfboy replied:
The point I'm trying to make is that if the `lms-main.scss` file just contains
```scss
@import 'overrides';
```
and that's all, then the output file will *not* have any of the core styles. If the theme developer wants to include the core styles and override some of them, then the `lms-main.scss` file should look like this:
```scss
@import 'overrides';
@import 'lms/static/sass/lms-main';
```
Does that make sense? Is there a clearer way I can phrase this in the documentation?

lamagnifica replied:
? I was talking about the "Next Steps" heading on line 190...

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,38 @@
+that is running Open edX. If you are using devstack or fullstack, you must
+be sure that the theme is present on the Vagrant virtual machine. If you
+are installing Open edX onto a server using the configuration scripts,
+you'll need to alter those scripts to install your theme onto the server,
```
you'll need to alter those scripts --> you must update the configuration scripts

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,38 @@
+theme on your server.
+
+Then, you will need to run Ansible, so that this variable is propagated to the
+``lms.envs.json`` file.
```
I'd invert this to something like,
To propagate this variable to the ``lms.envs.json`` file, you run Ansible.

lamagnifica commented on en_us/install_operations/source/theming/activate_theme.rst:
```diff
@@ -0,0 +1,38 @@
+
+Once the theme is installed on the server, you can activate it by setting
+the ``COMP_THEME_DIR`` Ansible variable to the full path of your theme's
+directory. You can do that by adding the following value to your
```
after the sentence ending "directory", what you've really got here is a set of discrete steps. Consider adding a procedure topic here and reformat the info.
===============
Activate a Theme
===============

To activate your theme, follow these steps. 

1. Open the (name) file in your text editor.
2. Edit the file to add the (name) parameter and the location of your theme.
3. To propagate this variable...
4. Restart the server...
5. To test the theme, ...

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,35 @@
+*************************
+
+In theory, all core assets in the LMS and Studio can be overriden by themed assets.
+In practice, most of the system assumes that core assets will *not* be overridden
```
will *not* be --> are not currently

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,35 @@
+In theory, all core assets in the LMS and Studio can be overriden by themed assets.
+In practice, most of the system assumes that core assets will *not* be overridden
+by themed assets, and you might see unexpected behavior if you override certain
+assets. So far, the list of supported themable assets is rather short:
```
is rather short: --> is limited to the following set.

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,35 @@
+
+.. list-table::
+   :header-rows: 1
+
```
add column widths (right now they're all the same width)

   :widths: 20 40 40

lamagnifica commented on en_us/install_operations/source/theming/supported_assets.rst:
```diff
@@ -0,0 +1,35 @@
+       ``images/profiles/default_120.png``
+       ``images/profiles/default_500.png``
+     - Default profile images for learner profiles. These are displayed in a
+       learner profile when the learner hasn't selected their own profile
```
These are displayed in a learner profile when the learner hasn't selected their own profile image. --> The default image displays for learners until they upload their own profile images. (or other rewrite to active voice, subject-verb agreement, and no contraction)

lamagnifica commented on en_us/install_operations/source/theming/supported_templates.rst:
```diff
@@ -0,0 +1,37 @@
+by themed templates. In practice, most of the system assumes that core templates
+will *not* be overridden by themed templates, and you might see unexpected
+behavior if you override certain templates.
+So far, the list of supported themable templates is rather short.
```
same

lamagnifica commented on en_us/install_operations/source/theming/supported_templates.rst:
```diff
@@ -0,0 +1,37 @@
+
+.. list-table::
+   :header-rows: 1
+
```
add col widths

lamagnifica commented on en_us/install_operations/source/theming/supported_templates.rst:
```diff
@@ -0,0 +1,37 @@
+     - ``header.html``
+     - The page header for the LMS. This template is used for the
+       top of the ``<body>`` element on the page, not for the ``<head>`` element
+       of the page. We recommend that you start by copying the contents of
```
We recommend --> EdX recommends (and on line 29)

lamagnifica commented on en_us/install_operations/source/theming/supported_templates.rst:
```diff
@@ -0,0 +1,37 @@
+   Currently, only templates loaded through the Mako templating system can be
+   overriden. This includes both Mako templates, and Underscore templates that
+   are loaded using the ``<%static:include%>`` Mako template tag.
+   Django templates currently cannot be overriden.
```
this note is a bit confusing right now. are there two Mako templates and an Underscore template? maybe:

Currently, you can only override templates that are loaded by the Mako templating system. You can override Mako templates and Underscore templates that are loaded using the ``<%static:include%>`` Mako template tag. You cannot override Django templates.

lamagnifica commented on en_us/install_operations/source/theming/sass.rst:
```diff
@@ -0,0 +1,194 @@
+
+.. note::
+
+    We suggest that you reference the `Sass/CSS Styleguide`_ that the edX
```
We suggest... --> The `Sass/CSS Styleguide`_, created by the edX UX (user experience) team, includes guidance for syntax preferences and suggestions for documenting variables.

lamagnifica commented:
Thanks again, @singingwolfboy for taking on this work. I've completed my copy edit. If you can make edits per my suggestions, that would be great, if not, one of the doc team can complete this and get it ready for publication.
The one file that's missing here is an entry in the change log (install_operations/source/front_matter/change_log). We typically add that just before publication.

lamagnifica commented:
@singingwolfboy did you intend to open a separate PR to add the change log entry?
